File: src/leetCode/c++/Solution542.cpp
//BFS O(mn) O(mn)
class Solution542 {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();
        vector<pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        queue<pair<int, int>> q;
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(mat[i][j] == 0){
                    q.push({i, j});
                }else{
                    mat[i][j] = -1;//表示要更新
                }
            }
        }

        //BFS只需要遍歷一次, 因為再次遍歷後的數字一定會較大, 並不會再次覆蓋
        while(!q.empty()){
            pair<int, int> p = q.front(); q.pop();

            for(auto dir : dirs){
                int x = p.first + dir.first;
                int y = p.second + dir.second;
                if(x < 0 || x >= m || y < 0 | y >= n || mat[x][y] != -1)
                    continue;

                mat[x][y] = mat[p.first][p.second] + 1;
                q.push({x, y});//相當於java的offer
            }
        }

        return mat;
    }
};

File: java/Solution542.java
package leetCode.java;

import java.util.*;

//BFS O(mn) O(mn)
class Solution542 {
    public int[][] updateMatrix(int[][] mat) {
        int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int m = mat.length;
        int n = mat[0].length;
        Queue<int[]> queue = new LinkedList<>();
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(mat[i][j] == 0){
                    queue.offer(new int[]{i, j});
                }else{
                    mat[i][j] = -1;//要更新的格子
                }
            }
        }

        while(!queue.isEmpty()){
            int[] posi = queue.poll();
            for(int[] dir : dirs){
                int i = posi[0] + dir[0];
                int j = posi[1] + dir[1];
                if(i < 0 || i >= m || j < 0 || j >= n || mat[i][j] != -1)
                    continue;

                mat[i][j] = mat[posi[0]][posi[1]] + 1;
                queue.offer(new int[]{i, j});
            }
        }
        return mat;
    }
}